tensorboard/compat/tensorflow_stub/io/gfile.py

```diff
@@ -401,6 +411,222 @@ def stat(self, filename):
                 raise
 
 
+class FSSpecFileSystem(object):
+    """Provides filesystem access via fsspec."""
+
+    SEPARATOR = "://"
+    CHAIN_SEPARATOR = "::"
```

Comment:
I definitely agree that this complicates the implementation a lot, but I wanted to chime in with a quick note that chaining is often used for wrapping a filesystem within a cache, i.e. `filecache::s3://bucket/key`. That's super useful when using remote filesystems with high latency, since it lets you use the cached copies for repeated reads/writes rather than making additional network requests. 

I wonder if there's some simple way to handle joining and globs that doesn't break with chaining?

Reply:
Ok, good to know that's desirable.

Maybe as a compromise it would be possible to restrict the chaining then to only allow the protocol separator in the final component of the chain?  So that would allow `filecache::s3://bucket/key` but not the more complicated cases from https://filesystem-spec.readthedocs.io/en/latest/features.html#url-chaining like `simplecache::zip://*.csv::gcs://bucket/afile.zip`.

That would resolve what is IMO the strangest thing, which is how joining and globbing inserts new path components into the middle of the path rather than appending them.  Effectively then we're just modeling it as though there's still only the single protocol separator, but one which potentially has a "procotol" composed of several filesystems chained together, which seems simpler to reason about.

Reply:
+1, I think that's a really good compromise as long as we can throw an exception with a nice message in cases that aren't handled by that.

Reply:
I think having chained file systems might be beneficial down the line but for now it seems unnecessarily complex. Always possible to relax the constraint later. Checking the URLs are valid w/ a nice exception seems doable

Reply:
added validation that only the last FS can have paths

Reply:
updated to handle `simplecache::/tmp/foo`